### PR TITLE
Apply clippy lints where they make sense

### PR DESCRIPTION
**Pull request by fuzzypixelz**
Quick run of `cargo clippy`.

This caught things like unnecessary `.clone()`s, and references. Though some lints suggested adding `Default` implementations for types that have a `new()` method, simply because "some users might need them", needless to say, I did not apply them as `nickel_lang` is not really a library with a public API.

### CODE REVIEW

**Comment by fuzzypixelz** [comment]
There is also somewhat of a false trigger of `needless_collect` in `merge`. It recommends delaying the collect of `contracts2` but that would lead to the iterator holding onto a mutable reference to `env`, which is used later on.

**Review by yannham — APPROVED** [review]

File: src/term.rs [hunk 1]
```diff
@@ -978,10 +978,7 @@ pub enum BinaryOp {
 
 impl BinaryOp {
     pub fn is_strict(&self) -> bool {
-        match self {
-            BinaryOp::Merge() => false,
-            _ => true,
-        }
+        !matches!(self, BinaryOp::Merge())
```
[the remark this answers is not in the document]
Comment (fuzzypixelz) [reply src/term.rs:981]: Macros also make compile times slower.

File: src/destruct.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ pub enum Match {
 pub enum LastMatch {
     /// The last field is a normal match. In this case the pattern is "closed" so every record
     /// fields should be matched.
-    Match(Match),
+    Match(Box<Match>),
```
[the remark this answers is not in the document]
Comment (fuzzypixelz) [reply src/destruct.rs:28]: Yes. Apparently the first variant is `416` bytes while the second one is only `40` bytes. An order of magnitude bigger :)

**Comment by fuzzypixelz** [comment]
@yannham Do we want to explicitly `#[allow(...)]` the lints we don't care about? Maybe even at the crate level? Because then we could integrate it in the CI using something like `cargo clippy -- -D warnings`

**Comment by yannham** [comment]
@fuzzypixelz yes, very good idea. This is what datered me from running clippy in the CI (too aggressive rules). Let's do that in a separate PR.

**Pull request merged**